File: my-folder/0238-product-of-array-except-self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int product = 1;
        int noOfZeros = 0;
        for (int i = 0; i < len; i++) {
            if (nums[i] != 0) {
                product = product * nums[i];
            } else {
                noOfZeros++;
            }
        }
        int[] arr = new int[len];
        for (int i = 0; i < len; i++) {
            if (nums[i] != 0) {
                if (noOfZeros == 0) {
                    arr[i] = product / nums[i];
                } else {
                    arr[i] = 0;
                }
            } else {
                if (noOfZeros == 1) {
                    arr[i] = product;
                } else {
                    arr[i] = 0;
                }
            }
        }
        return arr;
    }
}

File: my-folder/0238-product-of-array-except-self/solution.py
class Solution(object):
    def productExceptSelf(self, nums):
        contains_zero = False
        contains_multiple_zeroes = False
        total_product = 1
        for num in nums:
            if num == 0:
                if contains_zero:
                    contains_multiple_zeroes = True
                else:
                    contains_zero = True
            else:
                total_product *= num
        new_arr = []
        if contains_multiple_zeroes:
            return map(lambda x: 0, nums)
        for num in nums:
            if num == 0:
                new_arr.append(total_product)
            else:
                if contains_zero:
                    new_arr.append(0)
                else:
                    new_arr.append(total_product / num)
        return new_arr
            
        
